snap/component.go

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package snap
+
+import (
+	"fmt"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+)
+
+// ComponentInfo is the content of a component.yaml file.
+type ComponentInfo struct {
+	Component   ComponentId   `yaml:"component"`
+	Type        ComponentType `yaml:"type"`
+	Version     string        `yaml:"version"`
+	Summary     string        `yaml:"summary"`
+	Description string        `yaml:"description"`
+}
+
+// ComponentId contains the component name and the owner snap name.
+type ComponentId struct {
+	SnapName      string
+	ComponentName string
+}
+
+func (cid *ComponentId) UnmarshalYAML(unmarshall func(interface{}) error) error {
+	idStr := ""
+	if err := unmarshall(&idStr); err != nil {
+		return err
+	}
+
+	names := strings.Split(idStr, "+")
+	if len(names) != 2 {
+		return fmt.Errorf("incorrect component name %q", idStr)
+	}
+
+	*cid = ComponentId{SnapName: names[0], ComponentName: names[1]}
+
+	return nil
+}
+
+// ReadComponentInfoFromCont reads ComponentInfo from a snap component container.
+func ReadComponentInfoFromCont(compf Container) (*ComponentInfo, error) {
+	yamlData, err := compf.ReadFile("meta/component.yaml")
+	if err != nil {
+		return nil, err
+	}
+
+	var ci ComponentInfo
+
+	if err := yaml.UnmarshalStrict(yamlData, &ci); err != nil {
+		return nil, fmt.Errorf("cannot parse component.yaml: %s", err)
+	}
+
+	if err := ci.validate(); err != nil {
+		return nil, err
+	}
+
+	return &ci, nil
+}
+
+// FullName returns the full name of the component, which is composed
+// by snap name and component name.
+func (ci *ComponentInfo) FullName() string {
+	return fmt.Sprintf("%s+%s", ci.Component.SnapName, ci.Component.ComponentName)
+}
+
+// Validate performs some basic validations on component.yaml values.
+func (ci *ComponentInfo) validate() error {
+	if ci.Component.SnapName == "" {
+		return fmt.Errorf("snap name for component cannot be empty")
+	}
+	if ci.Component.ComponentName == "" {
+		return fmt.Errorf("component name cannot be empty")
+	}
+	for _, name := range []string{ci.Component.SnapName, ci.Component.ComponentName} {
+		if err := ValidateName(name); err != nil {
+			return err
+		}
+	}
+	if ci.Type == "" {
+		return fmt.Errorf("component type cannot be empty")
+	}
+	if err := ValidateVersion(ci.Version); err != nil {
+		return err
+	}
```

Comment:
we have methods to validate Description for snaps, mmh, I don't see a method to validate summary though, bit strange, we need to look into that

Reply:
so from what I could find review-tools has a max length of 128 on summary, and snapcraft itself of 78. I would snapd would need to go with the more generous one